tests/unit/s2n_ktls_test_utils_test.c

```diff
@@ -0,0 +1,579 @@
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_stuffer_pair io_pair = { 0 },
+                    s2n_ktls_io_stuffer_pair_free);
+            EXPECT_OK(s2n_test_init_ktls_io_stuffer(server, client, &io_pair));
+
+            size_t records_to_send = 5;
+            struct iovec send_msg_iov = { .iov_len = S2N_TEST_TO_SEND };
+            struct msghdr send_msg = { .msg_iov = &send_msg_iov, .msg_iovlen = 1 };
+            size_t total_sent = 0;
+            for (size_t i = 0; i < records_to_send; i++) {
+                /* increment test data ptr */
+                send_msg_iov.iov_base = test_data + total_sent;
+                send_msg.msg_control = &i;
+
+                /* sendmsg */
+                ssize_t bytes_written = s2n_test_ktls_sendmsg_io_stuffer(server->send_io_context, &send_msg);
+                EXPECT_EQUAL(bytes_written, S2N_TEST_TO_SEND);
+                total_sent += bytes_written;
+            }
+
+            /* confirm sent data */
+            EXPECT_OK(s2n_test_validate_data(&io_pair.client_in, test_data, total_sent));
+            /* validate `records_to_send` records were sent  */
+            EXPECT_EQUAL(s2n_stuffer_data_available(&io_pair.client_in.ancillary_buffer), records_to_send * S2N_TEST_KTLS_MOCK_HEADER_SIZE);
+            /* validate ancillary header */
+            for (size_t i = 0; i < records_to_send; i++) {
+                EXPECT_OK(s2n_test_validate_ancillary(&io_pair.client_in, i, S2N_TEST_TO_SEND));
+                /* consume the header in order to then validate the next header */
+                EXPECT_NOT_NULL(s2n_stuffer_raw_read(&io_pair.client_in.ancillary_buffer, S2N_TEST_KTLS_MOCK_HEADER_SIZE));
+            }
+
+            EXPECT_EQUAL(io_pair.client_in.send_recv_msg_invoked_count, records_to_send);
+        };
+
+        /* Attempt send and expect EAGAIN error  */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_stuffer_pair io_pair = { 0 },
+                    s2n_ktls_io_stuffer_pair_free);
+            EXPECT_OK(s2n_test_init_ktls_io_stuffer(server, client, &io_pair));
+            /* disable growable to simulate blocked/network buffer full */
+            io_pair.client_in.data_buffer.growable = false;
+
+            size_t to_send = 1;
+            struct iovec send_msg_iov = { .iov_base = test_data, .iov_len = to_send };
+            struct msghdr send_msg = { .msg_iov = &send_msg_iov, .msg_iovlen = 1, .msg_control = &test_record_type };
+            /* attempt sendmsg and expect EAGAIN */
+            EXPECT_EQUAL(s2n_test_ktls_sendmsg_io_stuffer(server->send_io_context, &send_msg), S2N_FAILURE);
+            EXPECT_EQUAL(errno, EAGAIN);
+
```

Comment:
When I'm testing blocking, I like to do it in a little loop. Just to make sure it stays properly blocked. I've been burned. 🥲